internal/exec/util/util.go

```diff
@@ -15,20 +15,81 @@
 package util
 
 import (
+	"errors"
+	"os"
 	"path/filepath"
 
 	"github.com/coreos/ignition/internal/log"
 	"github.com/coreos/ignition/internal/resource"
 )
 
+var (
+	errEscapedMountpoint = errors.New("Symlink traversal resulted in path outside of filesystem")
+)
+
 // Util encapsulates logging and destdir indirection for the util methods.
 type Util struct {
 	DestDir string // directory prefix to use in applying fs paths.
+	IsRoot  bool   // whether or not DestDir is the root filesystem
 	Fetcher resource.Fetcher
 	*log.Logger
 }
 
+// splitPath splits /a/b/c/d into [a, b, c, d]
+// golang-- for making me write this
+
+func splitPath(p string) []string {
+	dir, file := filepath.Split(p)
+	if dir == "" || dir == "/" {
+		return []string{file}
+	}
+	dir = filepath.Clean(dir)
+	return append(splitPath(dir), file)
+}
+
+func wantsToEscape(p string) bool {
+	return filepath.HasPrefix(filepath.Clean(p), "..")
+}
+
 // JoinPath returns a path into the context ala filepath.Join(d, args)
-func (u Util) JoinPath(path ...string) string {
-	return filepath.Join(u.DestDir, filepath.Join(path...))
+func (u Util) JoinPath(path ...string) (string, error) {
+	components := []string{}
+	for _, tmp := range path {
+		components = append(components, splitPath(tmp)...)
+	}
+
+	realpath := "/"
+	for _, component := range components {
+		tmp := filepath.Join(realpath, component)
+		s, err := os.Lstat(filepath.Join(u.DestDir, tmp))
+		if os.IsNotExist(err) {
+			realpath = tmp
+			continue
+		} else if err != nil {
+			return "", err
+		}
+
+		if s.Mode()&os.ModeSymlink == 0 {
+			realpath = tmp
+			continue
+		}
+
+		symlinkPath, err := os.Readlink(filepath.Join(u.DestDir, tmp))
+		if err != nil {
+			return "", err
+		}
+		if filepath.IsAbs(symlinkPath) {
+			if u.IsRoot {
+				realpath = "/"
+			} else {
+				return "", errEscapedMountpoint
+			}
+		}
+		if !u.IsRoot && wantsToEscape(symlinkPath) {
```

Comment:
Putting this in an `else` or `else if` of the `IsAbs` block might be clearer.

Also, it might be good to add a comment that in the `u.IsRoot` case, the link will still not be able to escape `u.DestDir`.